Add Enum.sum_by/2

sabiwara opened the pull request:
Proposal draft to include `Enum.sum/2`.

If we agree on the signature, will send a similar PR for `Enum.product/2`.

Side note: ~might also bench specific implementations for lists to see if it's worth it.~
Edit: [done](https://github.com/sabiwara/elixir_benches/commit/c5f0146e672468dc6e7bd4ff336483361b774c60), shows a 40% increase. Not bad, will update accordingly.

josevalim commented:
My biggest concern with `sum/2` and `product/2` is figuring out the API. We have `count/2`, but it works as a filter (understandably). `min` and `max` are also aggregations but the function configures the empty value. `sum_by` is not a good fit either, as it should behave closer to `frequencies_by` (which by the way it is not called `count_by`). So if we add this, each aggregation/2 function behaves slightly different and I am not sure if it will be confusing.

sabiwara commented:
I understand this concern, my thinking was that filtering is still possible by returning the noop value 0 or 1. Maybe we can explicitly provide an example doing so in the docs ?

An explicit alternative could be map_sum/2 , which somehow feels less elegant though.

kaaboaye commented:
> sum_by is not a good fit either, as it should behave closer to frequencies_by

Assuming that `frequencies_by` would have `value_fun` just like `group_by` `sum_by` wouldn't be so out of place.

<img width="863" alt="image" src="https://github.com/elixir-lang/elixir/assets/12039839/70e5c89e-1097-42b8-a393-30887711e915">

hauleth commented:
3rd time the charm? #11381 and #11455

sabiwara commented:
I've been thinking about this one, and `map_sum/2` would actually make a lot of sense if we think in terms of types.

It would be consistent with `map_join/2`, which has a almost the same signature and is the equivalent aggregation function for strings.

```elixir
map_join([t], (t -> String.t())) :: String.t()
map_sum([t], (t -> number())) :: number()
```

Type-wise, `_by` would be inconsistent (as you mentioned @josevalim) because `_by` functions usually return the original type, not the "mapped" type (or as keys in a map):

```elixir
max_by([t], (t -> u)) :: t
uniq_by([t], (t -> u)) :: [t]
```

Another benefit of `map_sum/2` is that the name is pretty clear and leaves no doubt about what it's doing.

whatyouhide commented:
I'm personally ok with `map_sum/2` as the name, what you just wrote makes sense @sabiwara. **r**. Also, we'd need `map_product/2`, just to point that out.

sabiwara commented:
> I’m not convinced we need this in the first place, though.

I think it would be good from the perspective of encouraging users to pick efficiency without sacrificing readability.
Most of the cases I've found myself needing `sum/1` are like the [one in the cheatsheet](https://hexdocs.pm/elixir/main/enum-cheat.html#sum-enum) and require mapping.
Right now one can be tempted to write `map(f) |> sum()` instead of a single reduce to make the intent clearer, yet this is not slightly but [highly more wasteful](https://github.com/sabiwara/iter/blob/v0.1.2/benchmarks/map_sum.results.txt).

sabiwara commented:
Closing for now since the appropriate solution still needs discussion.

PragTob commented:
@sabiwara you said closing as it still needs discussion but immediately reopened - accidental double click? :sweat_smile:

sabiwara commented:
> @sabiwara you said closing as it still needs discussion but immediately reopened - accidental double click? 😅

Nope, Jose asked me to keep it open 😉

josevalim commented:
Closing this, as we were unable to reach a conclusion.

aloisdg commented:
Learning Elixir right now. The absence of sum/2 is really sad.

 We are doing:
 
```elixir
      item[:quantity_by_size]
      |> Enum.map(fn {_k, v} -> v end)
      |> Enum.sum
```

and the exercise says

> Use the Enum.reduce function in the total_quantity function to practice. It's the best fitting Enum function for this task.

So are we doomed to rewrite a sum/2 with a low level tool like `reduce` each time? Cant we have something nicer?

Please let us write

```elixir
      item[:quantity_by_size]
      |> Enum.sum(fn {_k, v} -> v end)
```

FYI: Both C# has [sum/2](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.sum?view=net-8.0#system-linq-enumerable-sum-1(system-collections-generic-ienumerable((-0))-system-func((-0-system-nullable((system-int32)))))) and F# has [sum/2](https://fsharp.github.io/fsharp-core-docs/reference/fsharp-collections-listmodule.html#sumBy)

josevalim commented:
I have revisited my previous argument. And I found a flaw in it:

>  sum_by is not a good fit either, as it should behave closer to frequencies_by (which by the way it is not called count_by). 

We already have `max_by`, `dedup_by`, `chunk_by` and many others. And they all behave differently - and they should - because they are customizing different operations.

So if everyone agrees, I am fine with adding this as `sum_by` and `product_by`.

sabiwara commented:
@josevalim after thinking about this for a while, I was more leaning towards `map_sum/2` due to the very close situation with `map_join`. It could be caught by a linter rule which would be symmetrical with this [one](https://hexdocs.pm/credo/Credo.Check.Refactor.MapJoin.html), and type signatures are very similar too:
```
@spec map_join(Enumerable.t(x), x -> String.Chars.t()) :: String.t()
@spec map_sum(Enumerable.t(x), x -> number()) :: number()
```

OTOH, for a callback of type `x -> y`, `max_by` or `dedup_by` still return a `x` / `list(x)`.

I am fine either way, `sum_by/2` works as well, just wanted to share the rationale above.

sabiwara commented:
- Rebased
- Renamed to `sum_by` for now
- Added an entry to the cheatsheet and reworked the docs a bit

<img width="541" alt="Screenshot 2024-06-21 at 23 52 03" src="https://github.com/elixir-lang/elixir/assets/11598866/2cee8e2a-adba-4687-8ae5-b53447cef435">

wojtekmach commented on lib/elixir/lib/enum.ex:
```diff
@@ -3473,6 +3475,40 @@ defmodule Enum do
+      4
+
+  """
+  @doc since: "1.16.0"
```
```suggestion
  @doc since: "1.18.0"
```

:)

sabiwara replied:
Thanks great catch!

hauleth commented on lib/elixir/lib/enum.ex:
```diff
@@ -3473,6 +3475,40 @@ defmodule Enum do
+      iex> Enum.sum_by([], fn x -> x.count end)
+      0
+
+  Filtering can be achieved by returning `0` to remove elements:
```
```suggestion
  Filtering can be achieved by returning `0` to ignore elements:
```

I think that current wording may be a little bit misleading. I prefer such wording. Though I think that this whole paragraph could be removed as it "should be obvious".

sabiwara replied:
I like `ignore`, thank you for the suggestion.
I agree it should be obvious, yet I think it's worth it if it can help people think to use this style vs the much less efficient:

```elixir
[1, -2, 3] |> Enum.filter(fn x -> x > 0 end) |> Enum.sum()
```
